File: src/net/eduard/api/server/minigame/MinigameMap.java
package net.eduard.api.server.minigame;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import net.eduard.api.lib.storage.StorageAttributes;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.World;

import net.eduard.api.lib.modules.Mine;
import net.eduard.api.lib.game.Schematic;
import net.eduard.api.lib.modules.Copyable;
import net.eduard.api.lib.storage.Storable;

/**
 * Mapa da Sala
 * 
 * @author Eduard-PC
 *
 */
@StorageAttributes(indentificate = true)
public class MinigameMap implements Storable, Copyable {

	private String name;
	private String displayName;
	private String worldName;
	private int teamSize = 1;
	private int minPlayersAmount = 2;
	private int maxPlayersAmount = 20;
	private int neededPlayersAmount = 16;
	private int maxRounds=3;
	private Location spawn, lobby;
	private Map<String, Location> locations = new HashMap<>();
	private List<Schematic> bases = new ArrayList<>();
	private List<Location> spawns = new ArrayList<>();
	private Schematic map, feast;
	private Location feastLocation;

	public boolean isSolo() {
		return teamSize == 1;
	}

	public MinigameMap() {

		// TODO Auto-generated constructor stub
	}

	public MinigameMap copy() {
		return copy(this);

	}

	public MinigameMap(String name) {
		this.name = name;
		this.displayName = name;
	}

	public MinigameMap(Minigame minigame, String name) {
		this(name);
		minigame.getMaps().add(this);
	}

	@Override
	public Object restore(Map<String, Object> map) {
		return null;
	}

	@Override
	public void store(Map<String, Object> map, Object object) {
		// TODO Auto-generated method stub

	}

	public MinigameMap world(World world) {
		this.worldName = world.getName();
		for (Location spawn : this.spawns) {
			spawn.setWorld(world);
		}
		if (hasSpawn()) {
			spawn.setWorld(world);
		}
		if (hasLobby()) {
			lobby.setWorld(world);
		}
		for (Location loc : locations.values()) {
			loc.setWorld(world);
		}

		return this;
	}

	public void paste(Location relative) {
		world(relative.getWorld());

		Location dif = relative.clone().subtract(map.getRelative());
		if (hasSpawn()) {
			this.spawn.add(dif);
		}
		if (hasLobby()) {
			this.lobby.add(dif);
		}
		for (Location spawn : spawns) {
			spawn.add(dif);
		}
		map.paste(relative, true);

	}

	public Schematic getFeast() {
		return feast;
	}

	public void setFeast(Schematic feast) {
		this.feast = feast;
	}

	public boolean hasFeast() {
		// TODO Auto-generated method stub
		return feast != null;
	}

	public boolean hasLobby() {
		return lobby != null;
	}

	public boolean hasBases() {
		return !this.bases.isEmpty();
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public int getMinPlayersAmount() {
		return minPlayersAmount;
	}

	public void setMinPlayersAmount(int minPlayersAmount) {
		this.minPlayersAmount = minPlayersAmount;
	}

	public int getMaxPlayersAmount() {
		return maxPlayersAmount;
	}

	public void setMaxPlayersAmount(int maxPlayersAmount) {
		this.maxPlayersAmount = maxPlayersAmount;
	}

	public int getNeededPlayersAmount() {
		return neededPlayersAmount;
	}

	public void setNeededPlayersAmount(int neededPlayersAmount) {
		this.neededPlayersAmount = neededPlayersAmount;
	}

	public Location getSpawn() {
		return spawn;
	}

	public void setSpawn(Location spawn) {
		this.spawn = spawn;
	}

	public Location getLobby() {
		return lobby;
	}

	public void setLobby(Location lobby) {
		this.lobby = lobby;
	}

	public Schematic getMap() {
		return map == null ? (map = new Schematic()) : map;
	}

	public void setMap(Schematic map) {
		this.map = map;
	}

	public List<Location> getSpawns() {
		return spawns;
	}

	public void setSpawns(List<Location> spawns) {
		this.spawns = spawns;
	}

	public List<Schematic> getBases() {
		return bases;
	}

	public void setBases(List<Schematic> bases) {
		this.bases = bases;
	}

	public boolean hasSpawn() {
		return spawn != null;
	}

	public boolean hasSpawns() {
		return !spawns.isEmpty();
	}

	public boolean hasSchematic() {
		return this.map != null;
	}

	public World getWorld() {
		World mundo = Bukkit.getWorld(worldName);
		if (mundo == null) {
			mundo = Mine.loadWorld(worldName);

		}
		return mundo;
	}

	public MinigameMap fixWorld() {
		return world(getWorld());
	}

	public Location getFeastLocation() {
		return feastLocation;
	}

	public void setFeastLocation(Location feastLocation) {
		this.feastLocation = feastLocation;
	}

	public int getTeamSize() {
		return teamSize;
	}

	public void setTeamSize(int teamSize) {
		this.teamSize = teamSize;
	}

	public Map<String, Location> getLocations() {
		return locations;
	}

	public void setLocations(Map<String, Location> locations) {
		this.locations = locations;
	}

	public String getWorldName() {
		return worldName;
	}

	public void setWorldName(String worldName) {
		this.worldName = worldName;
	}

	public String getDisplayName() {
		return displayName;
	}

	public void setDisplayName(String displayName) {
		this.displayName = displayName;
	}

	public int getMaxRounds() {
		return maxRounds;
	}

	public void setMaxRounds(int maxRounds) {
		this.maxRounds = maxRounds;
	}

}

File: src/main/java/net/eduard/api/server/minigame/MinigameMap.kt
package net.eduard.api.server.minigame

import org.bukkit.Location
import org.bukkit.World

import net.eduard.api.lib.modules.Copyable
import net.eduard.api.lib.storage.annotations.StorageIndex
import org.bukkit.Bukkit
import org.bukkit.util.Vector

/**
 * Representa o Mapa da Sala do Minigame
 *
 * @author Eduard
 */


class MinigameMap(

    @StorageIndex
    var name: String = "mapa",
    var displayName: String = "mapinha"
) {
    @Transient
    lateinit var minigame: Minigame
    var feastRadius = 20
    var chestsMiniFeastLocations = mutableListOf<Location>()
    var chestsFeastsLocations = mutableListOf<Location>()
    var chestsIslandsLocations = mutableListOf<Location>()

    fun defaultWorldName(): String {
        return "${minigame.name}/map/$name"
    }


    var feastCenter: Location? = null
        get() {
            if (field == null)
                field = Location(world, 0.5, 100.5, 0.5)
            return field
        }


    var minPlayersAmount = 2
    var maxPlayersAmount = 20
    var neededPlayersAmount = 16
    var spawn: Location? = null
    var lobby: Location? = null
    var locations = mutableMapOf<String, Location>()
    var islands = mutableMapOf<Int, MinigameIsland>()
    var spawns = mutableListOf<Location>()
    val gameMap: MinigameSchematic?
        get() {
            return if (MinigameSchematic.hasSchematic(mapName()))
                MinigameSchematic.getSchematic(mapName())
            else null
        }

    val allChests get() = gameMap!!.getAllChests(feastCenter!!)
    val feastChests get() = allChests.filter { insideFeast(it.location) }

    fun mapName() = minigame.name + "-" + name
    val world: World
        get() = worldUsed.loadOrGet()
    var worldName
        get() = worldUsed.worldName
        set(value) {
            worldUsed.worldName = value
        }

    fun fixWorld() = world(world)
    val hasLobby get() = lobby != null
    val hasSpawn get() = spawn != null
    val hasSpawns get() = spawns.isNotEmpty()


    fun copyWorld(map: MinigameMap) {
        worldUsed.copy(map.worldName)
        fixWorld()
    }


    fun insideFeast(location: Location): Boolean {
        val raioDoFeast = (feastRadius * feastRadius)
        val distanciaDoBloco = feastCenter!!.distanceSquared(location)
        //  debug("Bloco: "+ location+" RaioDoFeast: "+ raioDoFeast+" distanciaBloco: " +distanciaDoBloco)
        return distanciaDoBloco < raioDoFeast
    }

    fun unloadWorld() = worldUsed.unload()

    fun resetWorld() {
        worldUsed.reset()
        fixWorld()
    }

    fun clearWorld() {
        worldUsed.clear()
        fixWorld()
    }

    fun copy(): MinigameMap {
        return Copyable.copyObject(this)
    }

    fun world(world: World): MinigameMap {
        this.worldName = world.name
        for (spawn in this.spawns) {
            spawn.world = world
        }
        if (hasSpawn) {
            spawn!!.world = world
        }
        if (hasLobby) {
            lobby!!.world = world
        }
        for (loc in locations.values) {
            loc.world = world
        }
        for (loc in chestsIslandsLocations) {
            loc.world = world
        }
        for (loc in chestsMiniFeastLocations) {
            loc.world = world
        }
        for (loc in chestsFeastsLocations) {
            loc.world = world
        }
        for ((_, island) in islands) {
            island.chest1Location?.world = world
            island.chest2Location?.world = world
            island.spawnLocation?.world = world
            island.centerLocation?.world = world
            island.lowLocation?.world = world
            island.highLocation?.world = world
            for (locChest in island.chestsLocation) {
                locChest.world = world
            }
        }
        feastCenter?.world = world

        return this
    }

    var worldUsed = MinigameWorld()
        get() {
            if (field.worldName == "minigame-map") {
                field.worldName = defaultWorldName()
            }
            return field
        }

    fun debug(msg: String) {
        Bukkit.getConsoleSender().sendMessage("§b[MinigameMap] §f$msg")
    }

    @Transient
    var preparedForPasting = false

    fun prepareForPasting(diference: Vector) {
        if (preparedForPasting){
            debug("Ja foi aplicado o diferencial nas locations.")
            return
        }
        debug("Aplicando diferencial em todas as locations")
        this.spawn?.add(diference)
        this.lobby?.add(diference)

        for (spawn in spawns) {
            spawn.add(diference)
        }
        for (chest in chestsIslandsLocations) {
            chest.add(diference)
        }
        for (chest in chestsFeastsLocations) {
            chest.add(diference)
        }

        for (chest in chestsMiniFeastLocations) {
            chest.add(diference)
        }
        for ((_, island) in islands) {
            island.chest1Location?.add(diference)
            island.chest2Location?.add(diference)
            island.spawnLocation?.add(diference)
            island.centerLocation?.add(diference)
            island.lowLocation?.add(diference)
            island.highLocation?.add(diference)
            for (locChest in island.chestsLocation) {
                locChest.add(diference)
            }
        }
        preparedForPasting = true
    }

    fun paste(): MinigameSchematic.SchematicPasting? {
        return paste(feastCenter!!)
    }

    fun paste(relative: Location): MinigameSchematic.SchematicPasting? {
        world(relative.world)
        feastCenter = relative
        val map = gameMap
        if (map == null) {
            debug("Nao foi encontrado o schematic do mapa " + mapName())
            return null
        }
        val difX = (relative.blockX - map.relative.blockX)
        val difY = (relative.blockY - map.relative.blockY)
        val difZ = (relative.blockZ - map.relative.blockZ)
        val diference = Vector(difX, difY, difZ)
        debug("Diferenca entre Ambos: $diference")
        prepareForPasting(diference)
        debug("Local central da colagem: " + relative.toVector())
        debug("Local central do Schematic: " + map.relative)
        return map.paste(relative, true)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as MinigameMap
        if (name != other.name) return false
        return true
    }

    override fun hashCode(): Int {
        return name.hashCode()
    }
}